datalad_dataverse/remote.py

```diff
@@ -92,111 +136,311 @@ def api(self):
+                   [p for p in self.files_latest.values()]
+
+    def _upload_file(self, datafile, key, local_file, remote_file):
+        """helper for both transfer-store methods"""
+        # If the remote path already exists, we need to replace rather than
+        # upload the file, since otherwise dataverse would rename the file on
+        # its end. However, this only concerns the latest version of the dataset
+        # (which is what we are pushing into)!
+        existing_id = [i for i, p in self.files_latest.items()
+                       if p == remote_file]
+        if existing_id:
+            self.message(f"Replacing {remote_file} ...", type='debug')
+            # Note: docstring of identifier appears to be incorrect. It's the id
+            # of the to be replaced file, not the dataset.
+            # TODO: PR to pyDataverse once confirmed.
+            response = self.api.replace_datafile(identifier=existing_id[0],
+                                                 filename=remote_file.name,
+                                                 json_str=datafile.json(),
+                                                 is_filepid=False)
+        else:
+            self.message(f"Uploading {remote_file} ...", type='debug')
+            response = self.api.upload_datafile(identifier=self.doi,
+                                                filename=local_file,
+                                                json_str=datafile.json())
+        response.raise_for_status()
+        uploaded_file = response.json()['data']['files'][0]
+        # update cache:
+        self.add_to_filelist(uploaded_file)
+        # remember dataverse's database id for this key
+        assigned_id = uploaded_file['dataFile']['id']
+        self.annex.setstate(key, assigned_id)
+
+    def _download_file(self, key, local_file, remote_file):
+        """helper for both transfer-retrieve methods"""
         data_api = DataAccessApi(
             base_url=self.url,
             # this relies on having established the NativeApi in prepare()
             api_token=self._token,
         )
-        dataset = self.api.get_dataset(identifier=self.doi)
-
+        recorded_id = self.annex.getstate(key)
+        if recorded_id:  # empty string otherwise
+            file_id = recorded_id
+        else:
+            available_id = [i for i, p in self.files_all.items()
+                            if p == remote_file]
+            if available_id:
+                file_id = available_id[0]
+            else:
+                # This shouldn't happen, since transfer_retrieve would only be
+                # called if checkpresent returned True before.
+                raise ValueError(f"File {remote_file} is unknown to remote")
+        response = data_api.get_datafile(file_id)
         # http error handling
-        dataset.raise_for_status()
+        response.raise_for_status()
+        with open(local_file, "wb") as f:
+            f.write(response.content)
```

Comment:
Would this approach be capable of downloading huge files? Or is this memory-constrained?